core/dispatcher/src/main/scala/whisk/core/invoker/Invoker.scala

```diff
@@ -232,6 +259,7 @@ class Invoker(
                 case None => {
                     activationCounter.next() // this is the global invoker counter
                     incrementUserActivationCounter(tran.msg.subject)
+                    tran.releaseResource(FailedActivation(transid))
```

Comment:
either we release the resource because of a FailedActivation, or, earlier when the container is put back. Note that in releaseResource, there is a one time flag that checks if the message was already sent. If the releaseResource call on line 262 succeeds it's because the activation failed to return the container (which generally means the future failed).